### ISSUE TITLE
Cannot use @Validated @ConfigurationProperties with Filters

### ISSUE THREAD

**Issue by OrangeDog** [spring-projects/spring-boot#8605]
Adding `@Validated` to an existing property-bound `Filter` causes it to fail.
```java
@Bean
@ConfigurationProperties("my-filter")
public Filter myFilter() {
    return new MyFilter();
}

@Validated
public class MyFilter extends OncePerRequestFilter {
    @NotEmpty @Getter @Setter
    private List<String> rules;
    // ....
}
```
```plain
[ERROR] [localhost-startStop-1] 14:12:27.361 [/] - Exception starting filter clientBlockingFilter
java.lang.NullPointerException: null
    at org.springframework.web.filter.GenericFilterBean.init(GenericFilterBean.java:176)
    at org.apache.catalina.core.ApplicationFilterConfig.initFilter(ApplicationFilterConfig.java:279)
    at org.apache.catalina.core.ApplicationFilterConfig.<init>(ApplicationFilterConfig.java:109)
    at org.apache.catalina.core.StandardContext.filterStart(StandardContext.java:4572)
    at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5215)
    at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
    at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1419)
    at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1409)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[ERROR] [localhost-startStop-1] 14:12:27.362 StandardContext - One or more Filters failed to start. Full details will be found in the appropriate container log file
[ERROR] [localhost-startStop-1] 14:12:27.362 StandardContext - Context [] startup failed due to previous errors
```
Presumably the workaround is to extract the properties to a separate class or (until #8024) remove `@Validated`.

**Comment by OrangeDog** [comment]
#8277 related?

**Comment by philwebb** [comment]
Trying to mix Spring concerns with `Filters` can cause lifecycle problems. Filters are loaded quite early so I imagine in this case they're being initialized before the configuration properties can be applied.

Confusingly the error looks like it's a NPE accessing a logger.

You could try using `DelegatingFilterProxyRegistrationBean` which should prevent early initialization.

**Comment by philwebb** [comment]
If you've got a sample application we can try, we can try to dig into this a little more?

**Comment by OrangeDog** [comment]
`DelegatingFilterProxyRegistrationBean` does seem to work, yes.

A sample application should be trivial. Just add a Filter bean as described to a web starter project.

**Comment by OrangeDog** [comment]
> Confusingly the error looks like it's a NPE accessing a logger.

Maybe because `getClass()` returns a JDK proxy?

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by OrangeDog** [comment]
I'm not sure what more is expected. This is literally the entire application.
```java
@SpringBootApplication
public class Application {

  public static void main(String[] args) {
    SpringApplication.run(Application.class, args);
  }

  @Bean
  @ConfigurationProperties("my-filter")
  public Filter myFilter() {
    return new MyFilter();
  }

  @Validated
  public static class MyFilter extends OncePerRequestFilter {
    @NotEmpty @Getter @Setter
    private List<String> rules;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
        throws ServletException, IOException
    {
      filterChain.doFilter(request, response);
    }
  }

}
```
```yaml
my-filter:
  rules: 'abc'
```

**Comment by OrangeDog** [comment]
Just an aside - the Application doesn't actually fail to start.
It continues after this error.
```plain
2017-03-22 11:58:53.880  INFO 9192 --- [           main] s.w.s.m.m.a.RequestMappingHandlerAdapter : Looking for @ControllerAdvice: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@501edcf1: startup date [Wed Mar 22 11:58:52 GMT 2017]; root of context hierarchy
2017-03-22 11:58:53.945  INFO 9192 --- [           main] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped "{[/error]}" onto public org.springframework.http.ResponseEntity<java.util.Map<java.lang.String, java.lang.Object>> org.springframework.boot.autoconfigure.web.BasicErrorController.error(javax.servlet.http.HttpServletRequest)
2017-03-22 11:58:53.946  INFO 9192 --- [           main] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped "{[/error],produces=[text/html]}" onto public org.springframework.web.servlet.ModelAndView org.springframework.boot.autoconfigure.web.BasicErrorController.errorHtml(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse)
2017-03-22 11:58:53.966  INFO 9192 --- [           main] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/webjars/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2017-03-22 11:58:53.966  INFO 9192 --- [           main] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2017-03-22 11:58:53.993  INFO 9192 --- [           main] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**/favicon.ico] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2017-03-22 11:58:54.084  INFO 9192 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Registering beans for JMX exposure on startup
2017-03-22 11:58:54.126  INFO 9192 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2017-03-22 11:58:54.129  INFO 9192 --- [           main] com.example.Application                  : Started Application in 2.386 seconds (JVM running for 2.703)
```

**Comment by wilkinsona** [comment]
This doesn't have anything specifically to do with Spring Boot. This test will fail with the same `NullPointerException`:

```java
package com.example;

import static org.mockito.Mockito.mock;

import java.io.IOException;

import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;

import org.junit.Test;
import org.springframework.validation.annotation.Validated;
import org.springframework.validation.beanvalidation.MethodValidationPostProcessor;
import org.springframework.web.filter.GenericFilterBean;

public class GenericFilterBeanProxyingTests {

	@Test
	public void filterInitWorksAfterProxying() throws ServletException {
		MethodValidationPostProcessor postProcessor = new MethodValidationPostProcessor();
		postProcessor.setProxyTargetClass(true);
		postProcessor.afterPropertiesSet();
		SomeFilter someFilter = new SomeFilter();
		someFilter = (SomeFilter) postProcessor.postProcessAfterInitialization(someFilter, "someFilter");
		someFilter.init(mock(FilterConfig.class));
	}

	@Validated
	static class SomeFilter extends GenericFilterBean {

		@Override
		public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
				throws IOException, ServletException {

		}

	}

}
```

The log output indicates the reason for the problem:

```
15:07:58.739 [main] INFO org.springframework.aop.framework.CglibAopProxy - Unable to proxy method [public final void org.springframework.web.filter.GenericFilterBean.init(javax.servlet.FilterConfig) throws javax.servlet.ServletException] because it is final: All calls to this method via a proxy will NOT be routed to the target instance.
```

The `init` method being `final` prevent calls to it from being routed to the target instance. This means that they're handled by the proxy instead. The proxy's fields are not initialized (because it doesn't call its super class's constructor). This means that `logger` is null and the NPE results.

In short, you can't use `@Validated` on a subclass of `GenericFilterBean` when `MethodValidationPostProcessor` is configured to generate CGLib proxies because the `init` method is final.

I'd recommend creating a separate POJO for your configuration properties rather than binding them straight to the filter. That will avoid this problem and also gives you a much better chance of getting sensible configuration metadata generated for it. You can then validate that POJO and use it in a `@Bean` method to create and configure your filter.

Alternatively, you could provide your own `MethodValidationPostProcessor` that is configured not to use CGLib.

**Comment by reggie-k** [comment]
Hi, I am facing the same issue, but couldn't find any decent explanation regaring how to use  DelegatingFilterProxyRegistrationBean  for this matter. Can someone please assist? USing SpringBoot 1.5.6

**Comment by wilkinsona** [comment]
@reggie-k If you need some assistance in addition to the explanations and suggestions above, please ask on Stack Overflow or Gitter.
